src/vs/editor/contrib/referenceSearch/browser/referenceSearch.ts

```diff
@@ -191,31 +191,33 @@ KeybindingsRegistry.registerCommandAndKeybindingRule({
 });
 
 
-export function provideReferences(model: editorCommon.IReadOnlyModel, position: Position): TPromise<Location[]> {
-
-	// collect references from all providers
-	const promises = ReferenceProviderRegistry.ordered(model).map(provider => {
-		return asWinJsPromise((token) => {
-			return provider.provideReferences(model, position, { includeDeclaration: true }, token);
-		}).then(result => {
-			if (Array.isArray(result)) {
-				return <Location[]>result;
-			}
-			return undefined;
-		}, err => {
-			onUnexpectedExternalError(err);
+export function provideReferences(model: editorCommon.IReadOnlyModel, position: Position): PPromise<void, Location[]> {
```

Comment:
Not a fan of `PPromise`. I'd suggest to leave it `TPromise<Location[]>` that resolves with *all* locations but have an optional argument `IProgess<T>`, as it's defined [here](https://github.com/Microsoft/vscode/blob/master/src/vs/platform/progress/common/progress.ts#L34). Something like this

```ts
export function provideReferences(
  model: editorCommon.IReadOnlyModel, 
  position: Position, 
  progress?: IProgress<Location[]>
): TPromise<Location[]> {

  // ...
}
```

The looks are nice but it's also closer to how this would be done with native promises. Also, less changes in unrelated files are required with this.